config/log.js

```diff
@@ -11,18 +12,26 @@
  */
 
 module.exports.log = {
-  /** *************************************************************************
-   *                                                                          *
-   * Valid `level` configs: i.e. the minimum log level to capture with        *
-   * sails.log.*()                                                            *
-   *                                                                          *
-   * The order of precedence for log levels from lowest to highest is:        *
-   * silly, verbose, info, debug, warn, error                                 *
-   *                                                                          *
-   * You may also set the level to "silent" to suppress all logs.             *
-   *                                                                          *
-   ************************************************************************** */
-
   level: 'info',
-  noShip: true, // don't display "fancy" Sails ship when starting the app
+  noShip: true,
+
+  // Custom formatter for production to handle multiline logs in Azure
+  custom:
+    process.env.NODE_ENV === 'production'
+      ? require('winston').createLogger({
+          format: require('winston').format.combine(
+            require('winston').format.timestamp(),
+            require('winston').format.printf(
+              ({ level, message, timestamp }) => {
+                const msg =
+                  typeof message === 'string'
+                    ? message
+                    : JSON.stringify(message);
+                return `${timestamp} [${level}] ${msg.replace(/\n/g, '\\n')}`;
+              }
+            )
+          ),
+          transports: [new (require('winston').transports.Console)()],
+        })
+      : undefined,
 };
```

Comment:
- `winston` can be required only once
- The `level` is already added by sails
- The `timestamp` is a great addition, it should also be present in dev

```suggestion
const winston = require('winston');

module.exports.log = {
  level: 'info',
  noShip: true,

  custom: winston.createLogger({
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.printf(({ level, message, timestamp }) => {
        let msg = message;
        if (process.env.NODE_ENV === 'production') {
          // Custom formatter for production to handle multiline logs in Azure
          msg = typeof message === 'string' ? message : JSON.stringify(message);
          msg = msg.replace(/\n/g, '\\n');
        }
        return `${timestamp} ${msg}`;
      })
    ),
    transports: [new winston.transports.Console()],
  }),
};
```

Reply:
You are absolutely right. Actually the timestamp is automatically added by the system in prod, I'll remove it from there and I'll add a custom logger for non-prod envs. to add the timestamp there.